### PR TITLE
Prevent users from submitting gcses as completed with an empty award year

### PR DESCRIPTION
**Pull request by JR-G**
## Context

Users are able to add a qualification in the 'GCSE or equivalent' section and mark it as completed without entering a value in the 'Year awarded' or 'Grade' section.

## Changes proposed in this pull request

For consistency with the incomplete degree logic, a flash message will now appear on the qualification review page, and prevent the session being redirected back to the application page, if they try to mark the qualification as complete and press continue:

|Before|After|
|------------|------------|
|Before|After|

It will also present a message on the application review page informing them it is incomplete, with a link back to the qualification review page:
|Before|After|
|------------|------------|
|Before |After |

## Guidance to review

Considering this is an edge case, is it acceptable for these messages to be generic rather than specifying which fields are incomplete?

## Link to Trello card

https://trello.com/c/iJKKl9KS

## Things to check

- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

**Review by gregknight1 — COMMENTED** [review]
Not sure what the numbers are for this edge case but it's preferable for us to be specific about the fields that need completing. Might be not worth it if it rarely happens?

**Comment by davidgisbey** [comment]
Hey Greg, i think it's a good point that this could be more granular. I've been thinking about a solution that we can use throughout the sections and am going to add a card to the tech debt column for the devs to discuss at the next tech debt meeting.

File: app/components/candidate_interface/gcse_qualification_review_component.rb [hunk 1]
```diff
@@ -33,6 +33,13 @@ def show_missing_banner?
       end
     end
 
+    def show_values_missing_banner?
+      if @submitting_application
+        gcse_completed = "#{@subject}_gcse_completed"
+        (!@application_form.send(gcse_completed) || @application_form.send("#{@subject}_gcse").incomplete_gcse_information?) && @editable && !@application_qualification.missing_qualification?
```
Comment (davidgisbey) [inline app/components/candidate_interface/gcse_qualification_review_component.rb:39]: Looking at this with fresh eyes. Can we pull this into a separate method so it's a bit clearer what is going on?

We can put this block

`(!@application_form.send(gcse_completed) || @application_form.send("#{@subject}_gcse").incomplete_gcse_information?)`

into a private method called `section_or_gcse_is_incomplete?` or something like that

File: app/models/application_qualification.rb [hunk 2]
```diff
@@ -65,6 +65,13 @@ def incomplete_degree_information?
     false
   end
 
+  def incomplete_gcse_information?
+    return true if award_year.nil?
+    return true if grade.nil? && structured_grades.nil?
```
Comment (davidgisbey) [inline app/models/application_qualification.rb:70]: I think we can just change combine these two lines, it's pretty clear what's going on.

So `return true if award_year.nil? || ( grade.nil? && structured_grades.nil?)`

Comment (davidgisbey) [reply app/models/application_qualification.rb:70]: I think we can also use the established pattern here of creating a constant 

EXPECTED_GCSE_DATA = %i[
    qualification_type
    subject
    award_year
  ].freeze

and then doing something like

`
  def incomplete_gcse_information?
     return true if grade.nil? && structured_grades.nil?
     return true if EXPECTED_GCSE_DATA.any? do |field_name|
      send(field_name).blank?
    end

    false
  end
`

File: app/components/candidate_interface/gcse_qualification_review_component.rb (hunk 1 above)
Comment (davidgisbey) [reply app/components/candidate_interface/gcse_qualification_review_component.rb:39]: On second thought we can also add this here too `!@application_qualification.missing_qualification?` as the method name covers the behaviour.

So 

```
private

def section_or_gcse_is_incomplete?
  (!@application_form.send(gcse_completed) || 
    @application_form.send("#{@subject}_gcse").incomplete_gcse_information?) && 
    !@application_qualification.missing_qualification?
end 

```

**Comment by gregknight1** [comment]
> Hey Greg, i think it's a good point that this could be more granular. I've been thinking about a solution that we can use throughout the sections and am going to add a card to the tech debt column for the devs to discuss at the next tech debt meeting.

Cool, that sounds like a plan @davidgisbey. Think we could probably improve the wording a bit too but will hold off until we know the way forward. Thanks.

File: app/components/candidate_interface/gcse_qualification_review_component.rb [hunk 3]
```diff
@@ -33,10 +33,21 @@ def show_missing_banner?
       end
     end
 
+    def show_values_missing_banner?
+      if @submitting_application
+        section_or_gcse_incomplete?
```
Comment (davidgisbey) [inline app/components/candidate_interface/gcse_qualification_review_component.rb:38]: can we move @editable to this line please?

**Review by davidgisbey — APPROVED** [review]
LGTM

**Pull request merged**